### PR TITLE
Use per-tuple memory context in the export SRF.

### PR DESCRIPTION
**Pull request by a1exsh**
The bottledwater_export() set returning function should allocate all
memory it needs to process a tuple in the dedicated context to avoid
filling up executor state or the SPI proc context (whichever happens
to be the current).

The result to be returned should be allocated on the said memory
context and it should be reset every time *before* producing output
for the next tuple.

### CODE REVIEW

**Comment by a1exsh** [comment]
[clabot:check]

**Comment by ept** [comment]
@a1exsh Terrific, thank you so much for tracking down the problem! This looks great.

**Comment by a1exsh** [comment]
You're welcome :-)

File: ext/snapshot.c [hunk 1]
```diff
@@ -145,7 +155,20 @@ Datum bottledwater_export(PG_FUNCTION_ARGS) {
             state->current_table++;
             if (state->current_table < state->num_tables) open_next_table(state);
         } else {
-            SRF_RETURN_NEXT(funcctx, PointerGetDatum(format_snapshot_row(state)));
+            /* SPI_cursor_fetch() leaves us in the SPI mem. context */
+            MemoryContextSwitchTo(state->memcontext);
+
+            /* clear any prior tuple result memory */
+            MemoryContextReset(state->memcontext);
```
Comment (a1exsh) [inline ext/snapshot.c]: In the hindsight, looks like we could just use `rsinfo->econtext->ecxt_per_tuple_memory` for that...

Comment (a1exsh) [inline ext/snapshot.c]: Scratch that: it doesn't work with "SELECT func()", only with "SELECT ... FROM func()".

**Pull request merged**